core/tests/performance_framework/selenium_driver.py

```diff
@@ -0,0 +1,221 @@
+            self.driver.get(page_url)
+
+        har_dict = self.proxy.har
+
+        self._clear_proxy()
+        self._clear_driver()
+
+        return har_dict
+
+    def get_page_session_timings(self, page_url):
+        """Retrieve page load timings for a page URL."""
+        # pylint: disable=unused-variable
+        with Xvfb() as xvfb:
+        # pylint: enable=unused-variable
+            self._setup_driver(use_proxy=False)
+
+            self.driver.get(page_url)
+
+            # To get entries for all the resources we can use:
+            #   driver.execute_script("return window.performance.getEntries();")
+            # To get only the timing we can use:
+            #   driver.execute_script("return window.performance.timing;")
+            page_session_timings = (
+                self.driver.execute_script("return window.performance"))
+
+        self._clear_driver()
+
+        return page_session_timings
+
+    def get_page_session_timings_cached_state(self, page_url):
+        """Retrieve page load timings for a page URL while simulating a
+        cached state i.e, a return user.
+        """
+        # pylint: disable=unused-variable
+        with Xvfb() as xvfb:
+        # pylint: enable=unused-variable
+            self._setup_driver(use_proxy=False)
+
+            # Fetch the page once. This leads to caching of various resources.
+            self.driver.get(page_url)
+
+            self.driver.get(page_url)
+
+            # To get entries for all the resources we can use:
+            #   driver.execute_script("return window.performance.getEntries();")
+            # To get only the timing we can use:
+            #   driver.execute_script("return window.performance.timing;")
+            page_session_timings = (
+                self.driver.execute_script("return window.performance"))
+
+        self._clear_driver()
+
+        return page_session_timings
+
+    def _setup_proxy(self, downstream_kbps=None, upstream_kbps=None,
+                     latency=None):
+        self.server = Server(self.browser_mob_path)
+        self.server.start()
+        self.proxy = self.server.create_proxy()
+
+        # The proxy server is pretty sluggish, setting the limits might not
+        # achieve the desired behavior.
+        if downstream_kbps and upstream_kbps and latency:
+            self.proxy.limits({
+                'downstream_kbps': downstream_kbps,
+                'upstream_kbps': upstream_kbps,
+                'latency': latency
+            })
+
+    def _setup_driver(self, use_proxy=False):
+        if self.browser == 'chrome':
+            chrome_options = webdriver.ChromeOptions()
+            # Disable several subsystems which run network requests in the
+            # background. This is for use when doing network performance
+            # testing to avoid noise in the measurements
+            chrome_options.add_argument("--disable-background-networking")
+
+            if use_proxy and self.proxy:
```

Comment:
Yeah, we are unnecessarily checking for `self.proxy`, the user should setup the proxy and then provide `use_proxy = true` as an argument.